src/python/pants/cache/s3_artifact_cache.py

```diff
@@ -0,0 +1,175 @@
+
+_NOT_FOUND = 0
+_NETWORK = 1
+_UNKNOWN = 2
+
+
+def _log_and_classify_error(e, verb, cache_key):
+  if _not_found_error(e):
+    logger.debug('Not Found During {0} {1}'.format(verb, cache_key))
+    return _NOT_FOUND
+  if _network_error(e):
+    logger.debug('Failed to {0} (network) {1}: {2}'.format(verb, cache_key, str(e)))
+    return _NETWORK
+  logger.debug('Failed to {0} (client) {1}: {2}'.format(verb, cache_key, str(e)))
+  return _UNKNOWN
+
+
+class S3ArtifactCache(ArtifactCache):
+  """An artifact cache that stores the artifacts on S3."""
+
+  def __init__(self, config_file, profile_name, artifact_root, s3_url, local):
+    """
+    :param artifact_root: The path under which cacheable products will be read/written
+    :param s3_url: URL of the form s3://bucket/path/to/store/artifacts
+    :param BaseLocalArtifactCache local: local cache instance for storing and creating artifacts
+    """
+    super(S3ArtifactCache, self).__init__(artifact_root)
+    url = urlparse(s3_url)
+    self._s3 = _connect_to_s3(config_file, profile_name)
+    self._path = url.path
+    if self._path.startswith('/'):
+      self._path = self._path[1:]
+    self._localcache = local
+    self._bucket = url.netloc
+
+  def try_insert(self, cache_key, paths):
+    logger.debug('Insert {0}'.format(cache_key))
+    # Delegate creation of artifacts to the local cache
+    with self._localcache.insert_paths(cache_key, paths) as tarfile:
+      with open(tarfile, 'rb') as infile:
+        # Upload artifact to the remote cache.
+        try:
+          response = self._get_object(cache_key).put(Body=infile)
+          response_status = response['ResponseMetadata']['HTTPStatusCode']
+          if response_status < 200 or response_status >= 300:
+            raise NonfatalArtifactCacheError('Failed to PUT (http error) {0}: {1}'.format(
+              cache_key, response_status))
+        except Exception as e:
+          raise NonfatalArtifactCacheError(
+            'Failed to PUT (core error) {0}: {1}'.format(cache_key, str(e)))
+
+  def has(self, cache_key):
+    logger.debug('Has {0}'.format(cache_key))
+    if self._localcache.has(cache_key):
+      return True
+    try:
+      self._get_object(cache_key).load()
+      return True
+    except Exception as e:
+      _log_and_classify_error(e, 'HEAD', cache_key)
+      return False
+
+  def use_cached_files(self, cache_key, results_dir=None):
+    logger.debug('GET {0}'.format(cache_key))
+    if self._localcache.has(cache_key):
+      return self._localcache.use_cached_files(cache_key, results_dir)
+
+    s3_object = self._get_object(cache_key)
+    try:
+      get_result = s3_object.get()
+    except Exception as e:
+      _log_and_classify_error(e, 'GET', cache_key)
+      return False
+
+    # Delegate storage and extraction to local cache
+    try:
+      return self._localcache.store_and_use_artifact(
+        cache_key, iter_content(get_result['Body']), results_dir)
```

Comment:
This needs to close `get_result['Body']`, otherwise python holds onto the file contents in memory.